api/src/repositories/search-index-respository.ts

```diff
@@ -0,0 +1,222 @@
+export type InsertNumberSearchableRecord = Pick<NumberSearchableRecord, InsertSearchableRecordKey>;
+export type InsertSpatialSearchableRecord = Pick<SpatialSearchableRecord, InsertSearchableRecordKey>;
+export type InsertStringSearchableRecord = Pick<StringSearchableRecord, InsertSearchableRecordKey>;
+
+/**
+ * A class for creating searchable records
+ */
+export class SearchIndexRepository extends BaseRepository {
+  /**
+   * Inserts a searchable datetime record.
+   *
+   * @param {InsertDatetimeSearchableRecord[]} datetimeRecords
+   * @return {*}  {Promise<DatetimeSearchableRecord[]>}
+   * @memberof SearchIndexRepository
+   */
+  async insertSearchableDatetimeRecords(
+    datetimeRecords: InsertDatetimeSearchableRecord[]
+  ): Promise<DatetimeSearchableRecord[]> {
+    defaultLog.debug({ label: 'insertSearchableDatetimeRecords' });
+
+    const queryBuilder = getKnex().queryBuilder().insert(datetimeRecords).into('search_datetime').returning('*');
+
+    const response = await this.connection.knex<DatetimeSearchableRecord>(queryBuilder);
+
+    if (response.rowCount !== datetimeRecords.length) {
+      throw new ApiExecuteSQLError('Failed to insert searchable datetime records', [
+        'SearchIndexRepository->insertSearchableDatetimeRecords',
+        'rowCount did not match number of supplied records to insert'
+      ]);
+    }
+
+    return response.rows;
+  }
+
+  /**
+   * Inserts a searchable number record.
+   *
+   * @param {InsertNumberSearchableRecord[]} numberRecords
+   * @return {*}  {Promise<NumberSearchableRecord[]>}
+   * @memberof SearchIndexRepository
+   */
+  async insertSearchableNumberRecords(
+    numberRecords: InsertNumberSearchableRecord[]
+  ): Promise<NumberSearchableRecord[]> {
+    defaultLog.debug({ label: 'insertSearchableNumberRecords' });
+
+    const queryBuilder = getKnex().queryBuilder().insert(numberRecords).into('search_number').returning('*');
+
+    const response = await this.connection.knex<NumberSearchableRecord>(queryBuilder);
+
+    if (response.rowCount !== numberRecords.length) {
+      throw new ApiExecuteSQLError('Failed to insert searchable number records', [
+        'SearchIndexRepository->insertSearchableNumberRecords',
+        'rowCount did not match number of supplied records to insert'
+      ]);
+    }
+
+    return response.rows;
+  }
+
+  /**
+   * Inserts a searchable spatial record.
+   *
+   * @param {InsertSpatialSearchableRecord[]} spatialRecords
+   * @return {*}  {Promise<SpatialSearchableRecord[]>}
+   * @memberof SearchIndexRepository
+   */
+  async insertSearchableSpatialRecords(
+    spatialRecords: InsertSpatialSearchableRecord[]
+  ): Promise<SpatialSearchableRecord[]> {
+    defaultLog.debug({ label: 'insertSearchableSpatialRecords' });
+
+    const queryBuilder = getKnex().queryBuilder().insert(spatialRecords).into('search_spatial').returning('*');
+
+    const response = await this.connection.knex<SpatialSearchableRecord>(queryBuilder);
```

Comment:
The generic type should be the second parameter in the function call: `this.connection.knex(queryBuilder, SpatialSearchableRecord);`

The way it is now, it just forces the result type without actually running the zod check.

Reply:
Good catch, fixed.